### PR TITLE
Add support for prebuilt asars

### PR DESCRIPTION
**Pull request by jsg2021**
* [x] I have read the [contribution documentation](https://github.com/electron-userland/electron-packager/blob/master/CONTRIBUTING.md) for this project.
* [x] I agree to follow the [code of conduct](https://github.com/electron/electron/blob/master/CODE_OF_CONDUCT.md) that this project follows, as appropriate.
* [x] The changes are appropriately documented (if applicable).
* [x] The changes have sufficient test coverage (if applicable).
* [x] The testsuite passes successfully on my local machine (if applicable).

**Summarize your changes:**
I implemented the --asar.filename option referenced in #161

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = true
+
+[*.md]
+trim_trailing_whitespace = false
+
+[package.json]
+indent_style = space
+indent_size = 2
```
Comment (malept) [inline .editorconfig]: If we're going to add this, it should be in a separate PR.

File: NEWS.md [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 [Unreleased]: https://github.com/electron-userland/electron-packager/compare/v12.0.0...master
 
+* Added `asar.filename` to support pre-built asar files.
```
Comment (malept) [inline NEWS.md]: Could you please adjust this to conform to the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format?

File: docs/api.md [hunk 3]
```diff
@@ -183,12 +183,14 @@ is not restricted to the official set if [`download.mirror`](#download) is set.
...
+- `filename` (*String*): circumvents the asar generation and lets you specify your own pre-built asar file/directory.
 - `ordering` (*String*): A path to an ordering file for packing files. An explanation can be found on the [Atom issue tracker](https://github.com/atom/atom/issues/10163).
 - `unpack` (*String*): A [glob expression](https://github.com/isaacs/minimatch#features), when specified, unpacks the file with matching names to the `app.asar.unpacked` directory.
 - `unpackDir` (*String*): Unpacks the dir to the `app.asar.unpacked` directory whose names exactly or pattern match this string. The `asar.unpackDir` is relative to `dir`.
 
   Some examples:
 
+  - `asar.filename = './dist/app.asar'` will short-circuit creating the asar, and just copy that one into the electron bundle.
```
Comment (malept) [inline docs/api.md]: I don't think there needs to be an example for this, it should be self-explanatory.

File: docs/api.md [hunk 4]
```diff
@@ -183,12 +183,14 @@ is not restricted to the official set if [`download.mirror`](#download) is set.
 *Boolean* or *Object* (default: `false`)
 
 Whether to package the application's source code into an archive, using [Electron's archive format](https://github.com/electron/asar). Reasons why you may want to enable this feature are described in [an application packaging tutorial in Electron's documentation](http://electron.atom.io/docs/v0.36.0/tutorial/application-packaging/). When the value is `true`, pass default configuration to the `asar` module. The configuration values listed below can be customized when the value is an `Object`. Supported parameters include, but are not limited to:
+- `filename` (*String*): circumvents the asar generation and lets you specify your own pre-built asar file/directory.
```
Comment (malept) [inline docs/api.md]: The description should start with "A path to", like `ordering` below it.

File: test/_setup.js [hunk 5]
```diff
@@ -67,13 +67,14 @@ function npmInstallForFixture (fixture) {
         return true
       } else {
         console.log(`Running npm install in fixtures/${fixture}...`)
-        return exec('npm install --no-bin-links', {cwd: fixtureDir})
+        return exec('npm install --no-bin-links --no-package-lock', {cwd: fixtureDir})
```
Comment (malept) [inline test/_setup.js]: This should be in a different PR.

File: test/asar.js [hunk 6]
```diff
@@ -53,3 +53,23 @@ util.testSinglePlatform('asar test', (t, opts) => {
...
+  return util.packageAndEnsureResourcesPath(t, opts)
+    .then(generatedResourcesPath => {
+      resourcesPath = generatedResourcesPath
+      return fs.stat(path.join(resourcesPath, 'app.asar'))
+    }).then(stats => {
+      t.true(stats.isFile(), 'app.asar should exist under the resources subdirectory when opts.asar.filename points to a prebuilt asar')
+      return fs.pathExists(path.join(resourcesPath, 'app'))
+    }).then(exists => {
+      return t.false(exists, 'app subdirectory should NOT exist when app.asar is built')
+    })
```
Comment (malept) [inline test/asar.js]: This test should check that the copied asar is the same as the input asar. There is a utility function to assert this, I believe.

File: test/fixtures/asar-prebuilt/package.json [hunk 7]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "PrebuiltAsar",
+  "version": "0.0.1",
+  "description": "A test of the renamed from `electron-prebuilt` to `electron`",
```
Comment (malept) [inline test/fixtures/asar-prebuilt/package.json]: The description needs to be updated.

File: platform.js [hunk 8]
```diff
@@ -154,6 +154,19 @@ class App {
 
     const dest = path.join(this.originalResourcesDir, 'app.asar')
     debug(`Running asar with the options ${JSON.stringify(asarOptions)}`)
+
+    if (asarOptions.filename) {
```
Comment (malept) [inline platform.js]: I think that if there are other `asarOptions` other than `filename` and `filename` is set, it should warn that the other options won't be used.

File: platform.js [hunk 9]
```diff
@@ -154,6 +154,19 @@ class App {
...
+      const src = path.resolve(asarOptions.filename)
+
+      return fs.stat(src)
+        .then(stat => {
+          const target = stat.isFile() ? dest : this.originalResourcesDir
+          debug(`Copying asar: ${src} to ${target}`)
+          return fs.copy(src, target, {overwrite: false, errorOnExist: true})
+        })
+        .then(() => fs.remove(this.originalResourcesAppDir))
```
Comment (malept) [inline platform.js]: I think for efficiency purposes, if `asar.filename` is set, we shouldn't even copy the app directory in the first place. Then we can get rid of this `fs.remove` call.

File: platform.js [hunk 10]
```diff
@@ -154,6 +154,19 @@ class App {
 
     const dest = path.join(this.originalResourcesDir, 'app.asar')
     debug(`Running asar with the options ${JSON.stringify(asarOptions)}`)
+
+    if (asarOptions.filename) {
+      const src = path.resolve(asarOptions.filename)
+
+      return fs.stat(src)
+        .then(stat => {
+          const target = stat.isFile() ? dest : this.originalResourcesDir
```
Comment (malept) [inline platform.js]: Could you explain how `target` is determined? Shouldn't target always be `dest`?

File: test/fixtures/asar-prebuilt/package.json [hunk 11]
```diff
@@ -0,0 +1,10 @@
...
+  "main": "main.js",
+  "devDependencies": {
+    "electron": "^1.8.0",
+    "electron-prebuilt": "^1.4.0"
```
Comment (malept) [inline test/fixtures/asar-prebuilt/package.json]: There should only be one item in `devDependencies`, and it should be `"electron": "1.3.1"`. Every time we add a new Electron version to the tests, it takes longer to run the tests due to downloading prebuilt binaries.

File: .editorconfig (hunk 1 above)
Comment (jsg2021) [reply .editorconfig]: sorry about that, I didn't mean to add that. I was trying to follow the current style and that file made my editor honor it.

File: platform.js (hunk 10 above)
Comment (jsg2021) [reply platform.js]: if the user specified a directory (containing their app.asar and additional resources) fs-extra's `copy` requires the dest to be a directory... if the src is a file, it requires dest to be a file.

File: platform.js (hunk 8 above)
Comment (jsg2021) [reply platform.js]: should that happen here? or in the argument parsing?

File: platform.js (hunk 9 above)
Comment (jsg2021) [reply platform.js]: I started to go down that route, but tried to keep the change minimal. I'll short circuit that.

File: platform.js (hunk 8 above)
Comment (malept) [reply platform.js]: It should happen here. If it happens in argument parsing, people who use the JS API (such as users of Electron Forge or ember-electron) will not get the warning.

File: platform.js (hunk 10 above)
Comment (malept) [reply platform.js]: It's not clear from the docs that you can specify a directory for `asar.filename`. Although even if it did, I would disagree with it because the name of the option makes that choice not obvious. I think the value of `asar.filename` should only be an ASAR file.

File: NEWS.md [hunk 12]
```diff
@@ -4,6 +4,10 @@
 
 [Unreleased]: https://github.com/electron-userland/electron-packager/compare/v12.0.0...master
 
+### Added
+
+* `asar.filename` to support pre-built asar files.
```
Comment (malept) [inline NEWS.md]: Minor nit, remove the period and add ` (#823)` at the end.

File: docs/api.md [hunk 13]
```diff
@@ -183,6 +183,7 @@ is not restricted to the official set if [`download.mirror`](#download) is set.
 *Boolean* or *Object* (default: `false`)
 
 Whether to package the application's source code into an archive, using [Electron's archive format](https://github.com/electron/asar). Reasons why you may want to enable this feature are described in [an application packaging tutorial in Electron's documentation](http://electron.atom.io/docs/v0.36.0/tutorial/application-packaging/). When the value is `true`, pass default configuration to the `asar` module. The configuration values listed below can be customized when the value is an `Object`. Supported parameters include, but are not limited to:
+- `filename` (*String*): A path to a pre-built asar file. This will circumvent the asar generation and lets you specify your own pre-built asar file.
```
Comment (malept) [inline docs/api.md]: This is kind of redundant. Let me think about a better way to word this.

File: platform.js [hunk 14]
```diff
@@ -99,26 +99,35 @@ class App {
...
-          this.originalResourcesAppDir,
-          this.opts.electronVersion,
-          this.opts.platform,
-          this.opts.arch
-        ])
-      } else {
-        return true
-      }
-    })
+    const asarOptions = common.createAsarOpts(this.opts) || {}
```
Comment (malept) [inline platform.js]: At this point `asarOptions` should be an instance variable, no need to calculate it twice.

File: platform.js [hunk 15]
```diff
@@ -99,26 +99,35 @@ class App {
...
+
+    // don't copy the app if the asar filename is specified.
+    const stage = (asarOptions.filename)
+      ? Promise.resolve()
+      : fs.copy(this.opts.dir, this.originalResourcesAppDir, {
+        filter: ignore.userIgnoreFilter(this.opts),
+        dereference: this.opts.derefSymlinks
+      })
```
Comment (malept) [inline platform.js]: I would rather have a real `if` statement than a multi-line ternary statement.

File: platform.js [hunk 16]
```diff
@@ -99,26 +99,35 @@ class App {
...
+          return hooks.promisifyHooks(this.opts.afterPrune, [
+            this.originalResourcesAppDir,
+            this.opts.electronVersion,
+            this.opts.platform,
+            this.opts.arch
+          ])
+        } else {
+          return true
+        }
+      })
```
Comment (malept) [inline platform.js]: I don't know whether afterPrune and afterCopy hooks should be run, since neither of those operations actually took place, and `this.originalResourcesAppDir` doesn't actually exist.

Comment (jsg2021) [reply platform.js]: Yeah, I didn't what those hooks were for... I've changed it to just return early.

File: test/asar.js [hunk 17]
```diff
@@ -53,3 +53,26 @@ util.testSinglePlatform('asar test', (t, opts) => {
...
+      return fs.stat(path.join(resourcesPath, 'app.asar'))
+    }).then(stats => {
+      t.true(stats.isFile(), 'app.asar should exist under the resources subdirectory when opts.asar.filename points to a prebuilt asar')
+      return util.areFilesEqual(opts.asar.filename, path.join(resourcesPath, 'app.asar'))
+  }).then(eql => {
+      t.true(eql, 'app.asar should equal the prebuilt asar')
+      return fs.pathExists(path.join(resourcesPath, 'app'))
+    }).then(exists => {
+      return t.false(exists, 'app subdirectory should NOT exist when app.asar is built')
+    })
```
Comment (malept) [inline test/asar.js]: This can be:

```js
      }).then(exists => t.false(exists, 'app subdirectory should NOT exist when app.asar is built'))
```

File: platform.js [hunk 18]
```diff
@@ -147,14 +153,31 @@ class App {
...
     const dest = path.join(this.originalResourcesDir, 'app.asar')
-    debug(`Running asar with the options ${JSON.stringify(asarOptions)}`)
-    return pify(asar.createPackageWithOptions)(this.originalResourcesAppDir, dest, asarOptions)
+    debug(`Running asar with the options ${JSON.stringify(this.asarOptions)}`)
+
+    if (this.asarOptions.filename) {
+      const src = path.resolve(this.asarOptions.filename)
+      if (Object.keys(this.asarOptions).length > 1) {
+        common.warning('asar.filename has been specified, all other asar options will be ignored')
+      }
```
Comment (malept) [inline platform.js]: This should have a test.

File: platform.js (hunk 16 above)
Comment (malept) [reply platform.js]: I don't think returning early is the right solution... we may have to adjust the hook signature so that it's either the path to the resources app directory, or the path to the `app.asar` file (which would be a breaking change).

@MarshallOfSound since you have dealt with a lot of code relating to the hooks, you might have some insight here.

File: platform.js (hunk 18 above)
Comment (jsg2021) [reply platform.js]: what do you mean? Test that it warns? or test that the other options are ignored? If the later, how should that be done?

Comment (malept) [reply platform.js]: Test that it warns. This is an example of doing that (although it should probably be genericized into a function):

https://github.com/electron-userland/electron-packager/blob/a7437b908e68e97b288173759d0538efe7798c82/test/mas.js#L18-L33

Comment (jsg2021) [reply platform.js]: done.

**Comment by malept** [comment]
I'm also a little worried about the drop in code coverage, but let's see what it looks like after the latest changes go through CI.

**Comment by jsg2021** [comment]
I updated the usage.txt.

File: platform.js [hunk 19]
```diff
@@ -147,14 +153,31 @@ class App {
...
+
+      return fs.stat(src)
+        .then(stat => {
+          if (!stat.isFile()) {
+            throw new Error(`${src} must be an asar file.`)
```
Comment (malept) [inline platform.js]: This needs a test.

Comment (jsg2021) [reply platform.js]: done

File: docs/api.md [hunk 20]
```diff
@@ -183,6 +183,7 @@ is not restricted to the official set if [`download.mirror`](#download) is set.
 *Boolean* or *Object* (default: `false`)
 
 Whether to package the application's source code into an archive, using [Electron's archive format](https://github.com/electron/asar). Reasons why you may want to enable this feature are described in [an application packaging tutorial in Electron's documentation](http://electron.atom.io/docs/v0.36.0/tutorial/application-packaging/). When the value is `true`, pass default configuration to the `asar` module. The configuration values listed below can be customized when the value is an `Object`. Supported parameters include, but are not limited to:
+- `filename` (*String*): A path to a pre-built asar file. (This will circumvent the asar generation)
```
Comment (malept) [inline docs/api.md]: > A path to a pre-built asar file (this will circumvent asar generation).

File: test/asar.js [hunk 21]
```diff
@@ -53,3 +54,40 @@ util.testSinglePlatform('asar test', (t, opts) => {
...
+  opts.asar = {
+    'unpack': '*.pac',
+    'unpackDir': 'dir_to_unpack',
+    'filename': path.join(opts.dir, 'app.asar')
+  }
+  let resourcesPath
+  sinon.spy(console, 'warn')
+  return util.packageAndEnsureResourcesPath(t, opts)
+    .then(generatedResourcesPath => {
+      console.warn.calledWithExactly('WARNING: asar.filename has been specified, all other asar options will be ignored')
```
Comment (malept) [inline test/asar.js]: Shouldn't this be wrapped in `t.true()`?

File: usage.txt [hunk 22]
```diff
@@ -25,6 +25,7 @@ asar               whether to package the source code within your app into an ar
                    --asar and its sub-properties simultaneously.
 
                    Properties supported:
+                   - filename: path to a pre-built asar file. (This will circumvent the asar generation)
```
Comment (malept) [inline usage.txt]: > path to a pre-built asar file (this will circumvent asar generation)

File: test/asar.js (hunk 21 above)
Comment (jsg2021) [reply test/asar.js]: sorry, i’ve been in jest land for a while, i assumed sinon made the assertion.  i’ll update.

File: usage.txt (hunk 22 above)
Comment (jsg2021) [reply usage.txt]: done

File: test/asar.js (hunk 21 above)
Comment (jsg2021) [reply test/asar.js]: done

File: docs/api.md (hunk 20 above)
Comment (jsg2021) [reply docs/api.md]: done

**Comment by jsg2021** [comment]
None of this is breaking existing configurations. Without adding the `asar.filename` option, nothing changes.

The whole point of this option is to take control of the app bundle generation... the hooks can be external to this once the filename is used... if something is needed to be done to the app package, a more general hook can be added?

**Comment by jsg2021** [comment]
the after prune hook still needs to be called since its unrelated to copy.

**Comment by MarshallOfSound** [comment]
RE the hooks issue I believe the operation should throw an error if `afterCopy` or `afterPune` hooks are provided and a prebuilt ASAR file is provided , these are incompatible options.

Also I don't think this option should go into the `options.asar` option.  That object is typically sent pretty much through to the ASAR module.  This goes nowhere near the ASAR module, I'd prefer this if it was something like `options.prebuiltAsar`.  This should also throw errors if that option is used in conjunction with any other ASAR option.

Also @malept we should just make forge error out instantly if this option is used 😄

**Comment by malept** [comment]
The docs should be updated to note the `asar`, `afterCopy`, and `afterPrune` incompatibilities.

**Comment by jsg2021** [comment]
I think I got all the suggestions implemented.

File: common.js [hunk 23]
```diff
@@ -43,6 +43,25 @@ function parseCLIArgs (argv) {
     args.out = null
   }
 
+  if (args.prebuiltAsar) {
+    if (args.asar) {
+      warning('--prebuiltAsar and --asar options are incompatible. Ignoring --asar')
+      delete args.asar
+    }
+    console.log('\n\n%s\n\n', JSON.stringify(args, null, 2))
```
Comment (malept) [inline common.js]: ✂️  debug line

File: common.js [hunk 24]
```diff
@@ -43,6 +43,25 @@ function parseCLIArgs (argv) {
...
+    if (!args.prune) {
+      warning('--prebuiltAsar and --no-prune are incompatible. Ignoring --no-prune')
+    }
+
+    if (args.ignore) {
+      warning('--prebuiltAsar and --ignore are incompatible. Ignoring --ignore')
+    }
+
+    if (!args['deref-symlinks']) {
```
Comment (malept) [inline common.js]: Should be `!args.derefSymlinks`

File: common.js [hunk 25]
```diff
@@ -43,6 +43,25 @@ function parseCLIArgs (argv) {
...
+      warning('--prebuiltAsar and --no-deref-symlinks are incompatible. Ignoring --no-deref-symlinks')
+    }
+  }
+
```
Comment (malept) [inline common.js]: This should be outside of the CLI argument parsing, as I mentioned yesterday, because it will not give warnings to consumers of the JavaScript API.

File: docs/api.md [hunk 26]
```diff
@@ -67,6 +67,10 @@ packager({
 })
 ```
 
+**NOTE:** `afterCopy` will not be called if `prebuiltAsar` is set.
+
+---
```
Comment (malept) [inline docs/api.md]: The lines are unnecessary.

File: docs/api.md [hunk 27]
```diff
@@ -327,6 +336,14 @@ Arbitrary combinations of individual platforms are also supported via a comma-de
 The non-`all` values correspond to the platform names used by [Electron releases]. This value
 is not restricted to the official set if [`download.mirror`](#download) is set.
 
+##### `prebuiltAsar`
+
+*String*: (default: `undefined`)
+
+A path to a pre-built asar file. 
+
+**NOTE:** This option circumvents many options such as `asar`, `afterCopy`, `afterPrune`, `ignore`, `prune`, or any option related to the copy processes. These options are for generating an app directory/asar with this tool. However, if a prebuilt asar file is provided, these options are not needed.
```
Comment (malept) [inline docs/api.md]: I'd get rid of `, or any option related to the copy processes`. Also, it's a little too verbose in general, but I need some time to think of a better way to word it.

File: platform.js [hunk 28]
```diff
@@ -101,30 +101,34 @@ class App {
 
   copyTemplate () {
     // don't copy the app if the asar filename is specified.
-    if ((this.asarOptions || {}).filename) {
+    if (this.opts.prebuiltAsar) {
+      if (this.opts.afterCopy) {
+        return Promise.reject(new Error('afterCopy is incompatible with prebuiltAsar'))
```
Comment (malept) [inline platform.js]: You should just be able to `throw new Error`.

File: docs/api.md (hunk 26 above)
Comment (jsg2021) [reply docs/api.md]: I thought it was a nice visual grouping. I'll drop them.

Comment (malept) [reply docs/api.md]: I think if we're going to do that, it should be in a separate PR.

File: platform.js (hunk 28 above)
Comment (jsg2021) [reply platform.js]: This isn't an async function, I was trying to make as little of a splash as possible. I hadn't looked to see if the callers setup to handle thrown exceptions and not just promise rejections.

**Comment by jsg2021** [comment]
I've updated to use your suggested initialize()

**Comment by jsg2021** [comment]
looks like a transient error caused the signing test to fail for one of the jobs... but the rest seem to be passing. I haven't looked further into it, since its outside the scope of my changes and it has been passing.

**Comment by jsg2021** [comment]
Anything I should change?

**Comment by jsg2021** [comment]
@malept Would you like anything else? or do you think this is good?

File: NEWS.md [hunk 29]
```diff
@@ -24,6 +24,10 @@
 
 * Upgraded `galactus` to `^0.2.1` to fix a bug with relative paths
 
+### Added
```
Comment (MarshallOfSound) [inline NEWS.md]: This is now in the wrong place

Comment (jsg2021) [reply NEWS.md]: ah, should be fixed now

**Comment by jsg2021** [comment]
I hope all is well? Is this still wanted?

**Comment by jsg2021** [comment]
Your cleanup looks way better than my changes. :) I didn't want to change too much without fully comprehending the codebase. Sorry if I misunderstood your requests. Hope this gets merged 👍

**Comment by jsg2021** [comment]
I've been using this branch to bundle my apps, the flag works well. Hope we can merge soon.

**Comment by jsg2021** [comment]
Is this blocked by something? Anything I could help with?

**Comment by jsg2021** [comment]
🎉
